libopenage/gamestate/market.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2017-2017 the openage authors. See copying.md for legal info.
+
+#include "player.h"
+#include "market.h"
+
+namespace openage {
+
+Market::Market() {
+	// the default prices when a game starts
+	this->base_prices[game_resource::wood] = 100;
+	this->base_prices[game_resource::food] = 100;
+	this->base_prices[game_resource::stone] = 130;
+}
+
+// Price calculation is documented at doc/reverse_engineering/market.md#prices
+
+bool Market::sell(Player &player, const game_resource res) {
+	double mult = this->getMultiplier(player, false);
+
+	if (player.deduct(res, MARKET_TRANSACTION_AMOUNT)) {
+		// if deduct was successful
+		// calc the gold received from selling MARKET_TRANSACTION_AMOUNT of res
+		double amount = this->base_prices.get(res) * mult;
+		player.receive(game_resource::gold, amount);
+		// decrease the price
+		this->base_prices[res] -= MARKET_PRICE_D;
+		if (this->base_prices.get(res) < MARKET_PRICE_MIN) {
+			this->base_prices[res] = MARKET_PRICE_MIN;
+		}
+		return true;
+	}
+	return false;
+}
+
+bool Market::buy(Player &player, const game_resource res) {
+	double mult = this->getMultiplier(player, true);
+
+	// calc the gold needed to buy MARKET_TRANSACTION_AMOUNT of res
+	double price = this->base_prices.get(res) * mult;
+	if (player.deduct(game_resource::gold, price)) {
+		// if deduct was successful
+		player.receive(res, MARKET_TRANSACTION_AMOUNT);
+		// increase the price
+		this->base_prices[res] += MARKET_PRICE_D;
+		if (this->base_prices.get(res) > MARKET_PRICE_MAX) {
+			this->base_prices[res] = MARKET_PRICE_MAX;
+		}
+		return true;
+	}
+	return false;
+}
+
+std::shared_ptr<ResourceBundle> Market::getBuyPrices(const Player &player) const {
+	return this->getPrices(player, true);
+}
+
+std::shared_ptr<ResourceBundle> Market::getSellPrices(const Player &player) const {
+	return this->getPrices(player, false);
+}
+
+std::shared_ptr<ResourceBundle> Market::getPrices(const Player &player, const bool is_buy) const {
```

Comment:
I'm not sure how useful it is to have the resource bundle as `shared_ptr`. Isn't it sufficient as a stack object?